### PR TITLE
Solo5: "spt" target, application manifest generation, multiple devices support

### PR DESCRIPTION
**Pull request by mato**
These are the `mirage` front-end changes for #992.

### CODE REVIEW

**Comment by dinosaure** [comment]
LGTM

File: lib/mirage.ml [hunk 1]
```diff
@@ -565,6 +565,42 @@ let unikernel_opam_name name target =
...
+  let open Codegen in
+  let file = Fpath.(v "manifest.json") in
+  with_output file (fun oc () ->
+      let fmt = Format.formatter_of_out_channel oc in
+      append fmt "{ \"version\": 1, \"devices\": [ %s ] }" s;
+      R.ok ())
+    "Solo5 application manifest file"
+
+let generate_manifest_c () =
+  let json = "manifest.json" in
```
Comment (cfcs) [inline lib/mirage.ml]: Why do we not put this in `_build/` as well?

**Comment by mato** [comment]
Seeing as I've done all the heavy lifting here, could someone else from @mirage/core @mirage/devs please figure out what to do about `block_of_file` so that we can get `mirage-skeleton` to compile? From my questions on Slack today:

I need some help with Functoria for mirage/mirage#993 - @samoht or @drup perhaps? already asked Hannes but he doesn't have a solution. (Thread follows) (edited) 

In order to get mirage-skeleton to compile with the new Solo5 changes, I need to pass a different value to block_of_file depending on whether or not we are configuring for Solo5.

This is point (2) mentioned here https://github.com/mirage/mirage/issues/992.

I've tried adding an `is_solo5` to `Mirage_key`, and so far so good:

```ocaml
val is_solo5: bool value
(** Is true iff the {!target} key is a Solo5-based target.
*)
```

```ocaml
let is_solo5 =
 Key.match_ Key.(value target) @@ function
 | #mode_solo5 -> true
 | #mode_xen | #mode_unix -> false
```

However, I can't figure out how to sensibly use this function in device-usage/block's config.ml. I've tried both of the following:

```
27 | let img = block_of_file (Key.(if_ is_solo5) "storage" "disk.img")
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type
         string Mirage.Key.value = string Functoria_key.value
       but an expression was expected of type string
```

and

```
27 | let img = block_of_file (if Key.is_solo5 then "storage" else "disk.img")
                                 ^^^^^^^^^^^^
Error: This expression has type
         bool Mirage.Key.value = bool Functoria_key.value
       but an expression was expected of type bool
       because it is in the condition of an if-statement
```

**Comment by samoht** [comment]
@mato You can write this:

```ocaml
let img = Key.(pure block_of_file $ if_ is_solo5 "storage" "disk.img")
```

But this will make `img` something of type `block impl value` and might need to be propagated to the various use points.

**Comment by samoht** [comment]
After looking at device-usage/block's config.ml, what you want is probably more something like this:

```
let img = Key.(if_impl is_solo5 (block_of_file "storage") (block_of_file "disk.img"))
```

**Comment by mato** [comment]
@samoht I added `is_solo5` in 741ea6d. Feel free to push to mirage/mirage-skeleton and see if you can get the CI to pass here.

**Comment by hannesm** [comment]
I took the initiative (here https://github.com/hannesm/mirage-skeleton/commit/1acfba506d1af4acd0a4670183e27081de27fc8a) and pushed a commit to this PR which uses that mirage-skeleton remote. let's see whether travis succeeds. if that is the case, I'll merge and cut a release!

**Comment by hannesm** [comment]
with the most recent force-push a mirage-skeleton compiles fine within CI (see https://travis-ci.org/mirage/mirage/builds/592619722 for result) - the only issue left is qubes backend and device-usage/prng failing to install solo5-bindings-spt -- which I'll ignore for now (and in a later step fix the prng unikernel). thanks for this work!

**Pull request merged**